pkg/controller/sparkapplication/controller.go

```diff
@@ -17,465 +17,506 @@ limitations under the License.
-		UpdateFunc: controller.onUpdate,
-		DeleteFunc: controller.onDelete,
+	// Watch for Pod created by SparkApplication
+	err = c.Watch(&source.Kind{Type: &apiv1.Pod{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &v1beta1.SparkApplication{},
 	})
-	controller.applicationLister = crdInformer.Lister()
-
-	podsInformer := podInformerFactory.Core().V1().Pods()
-	sparkPodEventHandler := newSparkPodEventHandler(controller.queue.AddRateLimited, controller.applicationLister)
-	podsInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
-		AddFunc:    sparkPodEventHandler.onPodAdded,
-		UpdateFunc: sparkPodEventHandler.onPodUpdated,
-		DeleteFunc: sparkPodEventHandler.onPodDeleted,
-	})
-	controller.podLister = podsInformer.Lister()
-
-	controller.cacheSynced = func() bool {
-		return crdInformer.Informer().HasSynced() && podsInformer.Informer().HasSynced()
+	if err != nil {
+		return err
 	}
 
-	return controller
+	return nil
 }
 
-// Start starts the Controller by registering a watcher for SparkApplication objects.
-func (c *Controller) Start(workers int, stopCh <-chan struct{}) error {
-	glog.Info("Starting the workers of the SparkApplication controller")
-	for i := 0; i < workers; i++ {
-		// runWorker will loop until "something bad" happens. Until will then rekick
-		// the worker after one second.
-		go wait.Until(c.runWorker, time.Second, stopCh)
+func newReconciler(mgr manager.Manager, metricsConfig *util.MetricConfig) reconcile.Reconciler {
+	reconciler := &ReconcileSparkApplication{
+		client:   mgr.GetClient(),
+		scheme:   mgr.GetScheme(),
+		recorder: mgr.GetEventRecorderFor("spark-operator"),
 	}
 
-	// Wait for all involved caches to be synced, before processing items from the queue is started.
-	if !cache.WaitForCacheSync(stopCh, c.cacheSynced) {
-		return fmt.Errorf("timed out waiting for cache to sync")
+	if metricsConfig != nil {
+		reconciler.metrics = newSparkAppMetrics(metricsConfig.MetricsPrefix, metricsConfig.MetricsLabels)
+		reconciler.metrics.registerMetrics()
 	}
-	return nil
+	return reconciler
 }
 
-// Stop stops the controller.
-func (c *Controller) Stop() {
-	glog.Info("Stopping the SparkApplication controller")
-	c.queue.ShutDown()
-}
+var _ reconcile.Reconciler = &ReconcileSparkApplication{}
 
-// Callback function called when a new SparkApplication object gets created.
-func (c *Controller) onAdd(obj interface{}) {
-	app := obj.(*v1beta1.SparkApplication)
-	v1beta1.SetSparkApplicationDefaults(app)
-	glog.Infof("SparkApplication %s/%s was added, enqueueing it for submission", app.Namespace, app.Name)
-	c.enqueue(app)
+// ReconcileSparkApplication reconciles a SparkApplication object
+type ReconcileSparkApplication struct {
+	client           client.Client
+	scheme           *runtime.Scheme
+	metrics          *sparkAppMetrics
+	recorder         record.EventRecorder
+	crdClient        crdclientset.Interface
+	ingressURLFormat string
+	sm               sync.Map
```

Comment:
Keep previous app object here. Ideally we should get rid of this if we can transition without noise.